src/prop/proof_cnf_stream.cpp

```diff
@@ -0,0 +1,1265 @@
+{
+  return d_blocked.contains(pfn);
+}
+
+std::shared_ptr<ProofNode> ProofCnfStream::getProofFor(Node f)
+{
+  return d_proof.getProofFor(f);
+}
+
+bool ProofCnfStream::hasProofFor(Node f)
+{
+  return d_proof.hasStep(f) || d_proof.hasGenerator(f);
+}
+
+std::string ProofCnfStream::identify() const { return "ProofCnfStream"; }
+
+void ProofCnfStream::convertAndAssert(TNode node,
+                                      bool negated,
+                                      bool removable,
+                                      ProofGenerator* pg)
+{
+  Trace("cnf") << "ProofCnfStream::convertAndAssert(" << node
+               << ", negated = " << (negated ? "true" : "false")
+               << ", removable = " << (removable ? "true" : "false") << ")\n";
+  d_removable = removable;
+  if (pg)
+  {
+    Trace("cnf") << "ProofCnfStream::convertAndAssert: pg: " << pg->identify()
+                 << "\n";
+    Node toJustify = negated ? node.notNode() : static_cast<Node>(node);
+    d_proof.addLazyStep(
+        toJustify, pg, true, "ProofCnfStream::convertAndAssert:cnf");
+  }
+  convertAndAssert(node, negated);
+  // process saved steps in buffer
+  const std::vector<std::pair<Node, ProofStep>>& steps = d_psb.getSteps();
+  for (const std::pair<Node, ProofStep>& step : steps)
+  {
+    d_proof.addStep(step.first, step.second);
+  }
+  d_psb.clear();
+}
+
+void ProofCnfStream::convertAndAssert(TNode node, bool negated)
+{
+  Trace("cnf") << "ProofCnfStream::convertAndAssert(" << node
+               << ", negated = " << (negated ? "true" : "false") << ")\n";
+  switch (node.getKind())
+  {
+    case kind::AND: convertAndAssertAnd(node, negated); break;
+    case kind::OR: convertAndAssertOr(node, negated); break;
+    case kind::XOR: convertAndAssertXor(node, negated); break;
+    case kind::IMPLIES: convertAndAssertImplies(node, negated); break;
+    case kind::ITE: convertAndAssertIte(node, negated); break;
+    case kind::NOT:
+    {
+      // track double negation elimination
+      if (negated)
+      {
+        d_proof.addStep(node[0],
+                        PfRule::MACRO_SR_PRED_TRANSFORM,
+                        {node.notNode()},
+                        {node[0]});
+        Trace("cnf") << "ProofCnfStream::convertAndAssert: "
+                        "MACRO_SR_PRED_TRANSFORM added norm "
+                     << node[0] << "\n";
+      }
+      convertAndAssert(node[0], !negated);
+      break;
+    }
+    case kind::EQUAL:
+      if (node[0].getType().isBoolean())
+      {
+        convertAndAssertIff(node, negated);
+        break;
+      }
+      CVC4_FALLTHROUGH;
+    default:
+    {
+      // negate
+      Node nnode = negated ? node.negate() : static_cast<Node>(node);
+      // Atoms
+      SatLiteral lit = toCNF(node, negated);
+      bool added = d_cnfStream.assertClause(nnode, lit);
+      if (negated && added && nnode != node.notNode())
+      {
+        d_proof.addStep(
```

Comment:
Can you explain whats going on here?

Reply:
It's tracking double negation elimination. I'll add a comment.